Source ingestion pipeline recipe (#12109)
MaterializeInc/materialize: doc/developer/platform/architecture-storage.md

## Source ingestion

In response to a CREATE SOURCE command, the STORAGE layer should instantiate a *source ingestion pipeline*.
This pipeline does not need to be a dataflow, but as we will see it is helpful to use timestamps to maintain consistency.

> The goal of the source ingestion pipeline is to produce data that can be appended to a `persist` collection.

Reviewer: it's unclear to me whether you mean a persist _shard_ or a storage _collection_ here (and elsewhere). sorry for the nitpick, but I think it's especially important that we use consistent language in our architecture docs to avoid confusion and those are the terms used in the storage and persist design docs

my guess from context is that this one is actually a storage collection, since below about it being the "output of the source ingestion pipeline" and that might be sharded (e.g. by kafka partition) and have renditions (e.g. if we change decoding/envelope logic)

Author: That's a good point. As long as "storage collection" is understood to mean "persistent collection" I can just use that!

The result of this collection is the primary user-facing output of the source ingestion pipeline.
There may be other auxiliary information, like the reclocking collection, that feeds in to health checks and user-facing diagnostics.

We want source ingestion to satisfy at least these three properties:

> 1. Definiteness: this should be provided by `persist`, and isn't meant to be something we can screw up.

Reviewer 2: Is "we" here "the rest of the STORAGE layer"?

Author: Yup. "Anyone reading this document". Didn't want to write "you".

Reviewer 2: What about when the persist team is reading the doc, is what I mean! :D Maybe it's clear enough though!

Author: Ah ok, sorry. "Anyone reading this document for the purposes of following the recipe". I shall change to "recipe implementors"

2. "Fault tolerance": the source should produce specific output even in the presence of unspecified i. failures and restarts, ii. concurrency with other source implementations.
3. "Bounded input reliance": the raw input data we have to keep around to provide these properties should be as limited as possible.

## A recipe for source ingestion

> What follows is a recipe for source ingestion that means to provide three properties:

Reviewer 2: ```suggestion
What follows is a recipe for source ingestion that means to provide the three properties listed above.
```

> In addition, the "recipe" aspect of this means to make is "simple" to determine that you have these properties.

Reviewer 2: ```suggestion
In addition, the "recipe" aspect of this means to make it "simple" to determine that you have these properties.
```

    The messages are emitted in an order that matches the address (the address never goes backwards).
    The source can be initialized at an address.
    For each address, the source produces the same message to all readers (or "NA" if it's gone).

2. We have access to the `persist` framework that can record evolving collections

    The framework can accept arbitrary `Append` statements and confirm when the are durably recorded.
    It handles contended writes, picking at most one winner, and providing a single consistent history of updates.

3. We have an evolving understanding of what should happen to our data from raw sources to `persist`.

    We want some amount of flexibility here, as we may be surprised by the requirements of source decoding, or new opportunities.
    The "understanding" should reflect deterministic computation.
    Non-determinism is something we can do with an advanced recipe (more `persist` stuff, in the spirit of reclocking, where we record non-deterministic choices).

### The recipe

There is some prep work to do first:

1.  Identify all "raw sources" that contribute to some intended output (e.g. kafka topics, metadata, schema registry).

> 2.  "Reclock" each input, by choosing and recording in `persist` a collection of source addresses that are monotonically increasing (do not go backwards, but needn't always go forward).

Reviewer: this one I'd guess actually is a persist shard since (I think?) there's no way for `remap` to have renditions and (I assume?) we won't need to horizontally shard it out

Author: I'm happy to stick with "storage collection" here. At least, I don't think there is a need to force it to be one shard, and I'm happy to let someone else determine the policy for that.

Author: Ah, or maybe it would be polite to have it be only one shard. I'm not sure it's needed for reclocking, but other instances of the "record non-determinism" pattern might need single-shard behavior to make sure one doesn't introduce cross-shard inconsistency (e.g. winning the append to one shard and losing to another). In any case, I'll dress up the language to distinguish things.

Reviewer: much appreciated!

I'm also curious if introducing the distinction offers a chance to talk more specifically about what should happen if the shard uppers are at different timestamps. e.g. this following line later in the doc

> To start, determine `t` the `min` write frontier in `persist` for all outputs and all state.

Author: Yup. I've just noted that this is possible. In this case, it should all be good, in that .. in principle you should be able to restart from any time `s <= t` and your output should match any persisted output. So, while you might "fail to write" some produced updates, they should be identical to what exists (and if not, we need to start a new rendition / file some bugs). I'll put some text to that effect (that e.g. there is an `assert` one should be able to put in place about pre-committed data).

Author: Added to the text:
```
    Both of these changes may end up being aimed at a `persist` shard that has already committed updates at that time.
    This can be because on restart not all `persist` shards had the same frontier, or because another storage instance is contending on writes.
    The writes *should* be identical, owing to the determinism of the pipeline so far.
    If they are not identical (an instance could assert they are on any failed write) then a new rendition is required (and perhaps a bug report, if this is unexpected).
```

Reviewer 2: I think maybe you have this text in a commit that hasn't been pushed up yet, Frank?

Author: I did, but intentionally. Was waiting to batch up responses.

    We should use times that roughly track "real time" so that they align with the times of other collections.
    The reclocked inputs are going to be treated as if their data "become available" at the first moment for which the data's address is covered by the reclocking collection.

>     Note that this step doesn't use `t` yet.

Reviewer 2: This is the first mention of `t`. Did something get reordered, maybe?

Author: I moved reclocking up in to "prep work". Should be fixed in the most recent commit.

    Each raw source should produce exactly its outputs at times greater or equal to `t`.

5. From time `t` onward, apply the source pipeline to produce at each time:

    a. Changes to push at persist to update the output.

    b. Changes to push at persist to update the state.

These are the whole instructions.
Clearly Step 5. is not fully specified.
The property it must have is that two independent pipelines should produce the same output and state updates.
One way to make this happen is to behave as if each message becomes available at its timestamp, and at each time we "immediately" retire all complete work.
For example, the first timestamp for which we have both transaction metadata and all supporting data messages, we would output those data messages.
Alternately, the first timestamp for which we have both transaction metadata and all supporting data messages, *and no incomplete prior transactions*, we would output those data messages.
The specific choice of what to do depends on our goals, but one attempts to do it as if executing each timestamp one at a time in sequence, and doing all the work that is possible at each timestamp.

### Outcomes

> The reason we follow the recipe is because many good things happen automatically.

Reviewer 2: ```suggestion
The reason we follow the recipe is because we uphold the desired three properties automatically.
```

1. *Definiteness.*

    We get definiteness from `persist`.

>     I don't think we are able to screw this up.

Reviewer 2: This is a bit repetitive with the intro of these properties; not sure if intentional!

Author: It is! I just re-read it, and happy with it. The text is more substantial here, and is meant to present the rationale for why each of the properties might be satisfied, a discussion we cannot have when they are first introduced.

Reviewer 2: Same! I just meant property (1) which had the same "I don't think we can screw this up" text in both places. The new revision has that text only down here, which seems great!

2. *Fault Tolerance.*

    Our choice of `t` is meant to provide a "consistent snapshot" of our source pipeline.
    It is a moment at which we know exactly where we were in input processing (indicated by the reclocking collection), what our "state" was, and what the output corresponded to.
    With this consistent snapshot, one can resume work and proceed as if without interruption. Multiple concurrent instances should also produce identical outputs, because of the consistent snapshot and deterministic execution (onward, from the reclocking input).

> 3. *Conciseness.*

Reviewer 2: Should this be "bounded input reliance" to match the property from the start?

Author: Yes probably!

    The source can be initialized at an address.
    For each address, the source produces the same message to all readers (or "NA" if it's gone).

2. We have access to the `persist` framework that can record evolving collections

    The framework accepts arbitrary `Append` statements to individual shards and confirm when and whether they are durably recorded.
    It handles contended writes, picking at most one winner, and providing a single consistent history of updates for each shard.
    A collection may be formed of multiple shards, which may not have the same durable frontiers.

3. We have an evolving understanding of what should happen to our data from raw sources to `persist`.

    We want some amount of flexibility here, as we may be surprised by the requirements of source decoding, or new opportunities.
    The "understanding" should reflect deterministic computation.
    Non-determinism is something we can do with an advanced recipe (more `persist` stuff, in the spirit of reclocking, where we record non-deterministic choices).

### The recipe

There is some prep work to do first:

> 1.  Identify all "raw sources" that contribute to some intended output (e.g. kafka topics, metadata, schema registry).

Reviewer 3: I think I used "_base_ source" in my "descriptive" docs, but "raw source" seems like a better name!

2.  "Reclock" each input, by choosing and recording in `persist` shards a collection of source addresses that are monotonically increasing (do not go backwards, but needn't always go forward).

>     We should use times that roughly track "real time" so that they align with the times of other collections.

Author: I think so. We should be able to use any target time on any basis, as far as I can tell. Perhaps I should have kept reclocking out of this, other than to say that we should reclock inputs so that all are in the same time domain, and we have the expectation that any time reached by one will soon be reached by the the others.

    The reclocked inputs are going to be treated as if their data "become available" at the first moment for which the data's address is covered by the reclocking collection.

    As suggested, this may include things like the schema registry that aren't conventional inputs, but nonetheless are external sources that influence the output we will produce.

> 2. Identify the "state" that a streaming implementation of the intended functionality would need to maintain.

Reviewer 3: another good example in this section would be the `UPSERT` state (if we want to start kafka consumers at non-0 offsets), might be easier to reason about as an example than the transaction metadata

>     For example, if we are matching up transaction metadata with the data it references, we may need to maintain some outstanding metadata records and some outstanding data records, until all in a transaction match.

Author: Yes, though I didn't mention it because I was smart but because that text already existed. It is a great argument for why one may not be able to naively reclock everything. Or perhaps that the reclocking should be memoized, and not re-performed for independent sources (if that accomplishes the same thing). 

I was hoping that the *state* would be pipeline local, as it is a deterministic function of the clocked input collections.

>     For example, if we are matching up transaction metadata with the data it references, we may need to maintain some outstanding metadata records and some outstanding data records, until all in a transaction match.

Reviewer 4: > I was hoping that the state would be pipeline local, as it is a deterministic function of the clocked input collections.

yes, definitely agree with this!

    If we intend to hold back transactions until all strictly prior transactions have completed, we may need to hold on even to some completed transactions, awaiting the completion of the prior work.

3. Create `persist`-backed storage collections for the output and the identified "state".

With this prep work done, the assembly instructions are:

1. To start, determine `t` the `min` write frontier in `persist` for all shards of outputs and all state.

    The time `t` may be strictly less than some of the shards, which will mean that we cannot necessarily write at `t` for all shards.
    This is fine, and will be discussed in just a moment.

2. From time `t` onward, rehydrate the "state" associated with the source pipeline.

3. Restart each raw source from the addresses found in the persisted reclocking collection at `t`.

> 4. From time `t` onward, continually produce reclocked messages: messages from the raw source whose addresses are translated to timestamps using reclocking information from `persist`.

Reviewer 4: hmm do you think something like the current debezium transaction implementation will cause issues here: currently the logic timestamp comes from the metadata information -- the timestamp assigned to each data message is essentially ignored.  If we do the same thing here, we're just completely ignoring the persisted reclocking collection

Author: My hunch is that this will fight with the current DBZ approach. We might need to put all the constraints on the table, and now might be a good time to pre-think that before concluding that the pipeline is generally valid. For example, if we want to reclock data sources and then use their times to (in part) constrain the release of complete transactions, this makes it harder to emit completed transactions in order (each source would need to be aware of incomplete transactions across all tables, not just their own).

Repurposing the current DBZ pipeline would mean (I think) putting the consistency topic in charge of the reclocking of the data inputs (it would determine when certain data "are visible"), and it .. might do weird things like announce data as available before it is actually present in the topic. Or, .. it would need to somehow play the data out while looking at the transaction timestamps in order to determine the reclocking association.

Another take could be that this would require a reclocking of the reclocked data, into a timeline associated with the transactions.

The reasons we reclock, btw, aren't just progress reporting. It's there so that we can resume on a consistent cut through the inputs, and know how much or little of each input we need to retain / where to start from on a reboot. If we don't have this written down some way, it's a lot more custom to figure out how to recover correctly, and from where, and what of the input needs to be recorded to make this happen.

Reviewer 4: > Another take could be that this would require a reclocking of the reclocked data, into a timeline associated with the transactions.

Doing this conceptually checks out to me -- especially given your reminder about needing the reclocking for resumption purposes.  That the transaction time domain corresponds to the metadata topic time domain would be incidental but not strictly necessary for correctness of progress / resumption

    It is important that this be based on reclocking information we hear back from `persist`, rather than what we optimistically proposed at `persist`.
    This ensures that the right thing happens in the case there are contending implementations.

    Each raw source should now produce exactly its outputs at times greater or equal to `t`.

5. From time `t` onward, apply the source pipeline to produce at each time:

    a. Changes to push at persist to update the output.

>     b. Changes to push at persist to update the state.

Reviewer 3: the difference between the output/state persist shards and the "reclocking information we hear from `persist` is a bit unclear in this document. it refers to "reclocking collection" above, but in step 4 just says "reclocking information"

    Both of these changes may end up being aimed at a `persist` shard that has already committed updates at that time.
    This can be because on restart not all `persist` shards had the same frontier, or because another storage instance is contending on writes.

>     The writes *should* be identical, owing to the determinism of the pipeline so far.

Reviewer 3: so, in the case of sources like kinesis, where we can ensure we are past `t` when we restart, but multiple instances could produce non-identical data, do we get this by using a single instance?

    If they are not identical (an instance could assert they are on any failed write) then a new rendition is required (and perhaps a bug report, if this is unexpected).

These are all of the instructions.
Clearly Step 5. is not fully specified.
The property it must have is that two independent pipelines should produce the same output and state updates.
One way to make this happen is to behave as if each message becomes available at its timestamp, and at each time we "immediately" retire all complete work.
For example, the first timestamp for which we have both transaction metadata and all supporting data messages, we would output those data messages.

> Alternately, the first timestamp for which we have both transaction metadata and all supporting data messages, *and no incomplete prior transactions*, we would output those data messages.

Author: The only intent here was to call out that one may want more complicated rules, in order to produce transactional updates that appear to happen in the same timestamp order as input order. I'm not actually sure if that means each data message is a transaction (probably not?), but just that source pipeline implementors are welcome to use crazy rules as long as they are rules.

We want source ingestion to satisfy at least these three properties:

1. Definiteness: this property is provided by `persist`.

> 2. "Fault tolerance": The source pipeline should produce equivalent `persist` output when run when run with arbitrary unclean shutdowns and restarts, and arbitrary concurrent instances, as when it is run uninterrupted".

Reviewer 3: ```suggestion
2. "Fault tolerance": The source pipeline should produce equivalent `persist` output when run when run with arbitrary unclean shutdowns and restarts and/or arbitrary concurrent instances, as when it is run uninterrupted".
```
